snuba/web/query.py

```diff
@@ -209,33 +268,69 @@ def log_query_and_update_stats(
     """
 
     stats.update(query_settings)
+
+    query_metadata.query_list.append(
+        ClickhouseQueryMetadata(
+            sql=sql, timing=timer.for_json(), stats=stats, status=status, trace_id=trace_id,
+        )
+    )
+
+    return stats
+
+
+def record_query(
+    request: Request, timer: Timer, query_metadata: SnubaQueryMetadata
+) -> None:
     if settings.RECORD_QUERIES:
         # send to redis
-        state.record_query(
-            {
-                "request": request.body,
-                "sql": sql,
-                "timing": timer,
-                "stats": stats,
-                "status": status,
-            }
-        )
+        query_metadata.http_referrer = http_request.referrer
```

Comment:
Why is this bound so late here, rather at near the instantiation site?

This may also break or have confusing behavior (not sure specifically how Flask deals with this) if this setting is turned on for subscriptions, since there is no HTTP request when the subscription is being executed. The request that is being passed as the first parameter should already have the correct referrer bound, regardless of how the request was instantiated.

Reply:
I just checked this out of curiosity, and tests do fail if this setting is turned on with this error:

```
Traceback (most recent call last):
  File "/Users/ted/workspaces/snuba/tests/subscriptions/test_subscription.py", line 108, in test
    identifier = creator.create(subscription, Mock())
  File "/Users/ted/workspaces/snuba/snuba/subscriptions/subscription.py", line 33, in create
    parse_and_run_query(self.dataset, request, timer)
  File "/Users/ted/workspaces/snuba/snuba/web/query.py", line 320, in parse_and_run_query
    record_query(request_copy, timer, query_metadata)
  File "/Users/ted/workspaces/snuba/snuba/web/query.py", line 286, in record_query
    query_metadata.http_referrer = http_request.referrer
  File "/Users/ted/.virtualenvs/snuba/lib/python3.7/site-packages/werkzeug/local.py", line 348, in __getattr__
    return getattr(self._get_current_object(), name)
  File "/Users/ted/.virtualenvs/snuba/lib/python3.7/site-packages/werkzeug/local.py", line 307, in _get_current_object
    return self.__local()
  File "/Users/ted/.virtualenvs/snuba/lib/python3.7/site-packages/flask/globals.py", line 37, in _lookup_req_object
    raise RuntimeError(_request_ctx_err_msg)
RuntimeError: Working outside of request context.

This typically means that you attempted to use functionality that needed
an active HTTP request.  Consult the documentation on testing for
information about how to avoid this problem.
```

We should probably consider turning `RECORD_QUERIES` on in settings_test.py since it has the potential to be subtly problematic.